panda-rs/src/api/regs.rs

```diff
@@ -0,0 +1,212 @@
+#[cfg(feature = "x86_64")]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, EnumString, EnumIter)]
+pub enum Reg {
+    RAX = 0,
+    RCX = 1,
+    RDX = 2,
+    RBX = 3,
+    RSP = 4,
+    RBP = 5,
+    RSI = 6,
+    RDI = 7,
+    R8 = 8,
+    R9 = 9,
+    R10 = 10,
+    R11 = 11,
+    R12 = 12,
+    R13 = 13,
+    R14 = 14,
+    R15 = 15,
+}
+
+#[cfg(feature = "arm")]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, EnumString, EnumIter)]
+pub enum Reg {
+    R0 = 0,
+    R1 = 1,
+    R2 = 2,
+    R3 = 3,
+    R4 = 4,
+    R5 = 5,
+    R6 = 6,
+    R7 = 7,
+    R8 = 8,
+    R9 = 9,
+    R10 = 10,
+    R11 = 11,
+    R12 = 12,
+    LR = 13,
+    SP = 14,
+    IP = 15,
+}
+
+#[cfg(any(feature = "mips", feature = "mipsel"))]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, EnumString, EnumIter)]
+pub enum Reg {
+    ZERO = 0,
+    AT = 1,
+    V0 = 2,
+    V1 = 3,
+    A0 = 4,
+    A1 = 5,
+    A2 = 6,
+    A3 = 7,
+    T0 = 8,
+    T1 = 9,
+    T2 = 10,
+    T3 = 11,
+    T4 = 12,
+    T5 = 13,
+    T6 = 14,
+    T7 = 15,
+    S0 = 16,
+    S1 = 17,
+    S2 = 18,
+    S3 = 19,
+    S4 = 20,
+    S5 = 21,
+    S6 = 22,
+    S7 = 23,
+    T8 = 24,
+    T9 = 25,
+    K0 = 26,
+    K1 = 27,
+    GP = 28,
+    SP = 29,
+    FP = 30,
+    RA = 31,
+}
+
+// TODO: reg map
+//#[cfg(feature = "aarch64")]
+//#[derive(Debug, PartialEq, Eq, EnumString, EnumIter)]
+
+// TODO: reg map
+//#[cfg(feature = "ppc")]
+//#[derive(Debug, PartialEq, Eq, EnumString, EnumIter)]
+
+// Getters/setters -----------------------------------------------------------------------------------------------------
+
+/// Get stack pointer register
+pub fn reg_sp() -> Reg {
+
+    #[cfg(feature = "i386")]
+    return Reg::ESP;
+
+    #[cfg(feature = "x86_64")]
+    return Reg::RSP;
+
+    #[cfg(any(feature = "arm", feature = "mips", feature = "mipsel"))]
+    return Reg::SP;
+}
+
+/// Get return value register
+#[cfg(any(feature = "i386", feature = "x86_64"))]
+pub fn reg_ret_val() -> Reg {
+
+    #[cfg(feature = "i386")]
+    return Reg::EAX;
+
+    #[cfg(feature = "x86_64")]
+    return Reg::RAX;
+}
+
+/// Get return value registers
+/// Note that most C code will only use the first register, e.g. index 0 in returned `Vec`
+#[cfg(any(feature = "arm", feature = "mips", feature = "mipsel"))]
+pub fn reg_ret_val() -> Vec<Reg> {
+
+    #[cfg(feature = "arm")]
+    return vec![Reg::R0, Reg::R1, Reg::R2, Reg::R3];
+
+    #[cfg(any(feature = "mips", feature = "mipsel"))]
+    return vec![Reg::V0, Reg::V1];
+}
+
+/// Get return address register
+pub fn reg_ret_addr() -> Option<Reg> {
+
+    #[cfg(feature = "i386")]
+    return None;
+
+    #[cfg(feature = "x86_64")]
+    return None;
+
+    #[cfg(feature = "arm")]
+    return Some(Reg::LR);
+
+    #[cfg(any(feature = "mips", feature = "mipsel"))]
+    return Some(Reg::RA);
+}
+
+/// Read the current value of a register
+#[cfg(any(feature = "i386", feature = "x86_64", feature = "arm"))]
+pub fn get_reg(cpu: &CPUState, reg: Reg) -> target_ulong {
```

Comment:
For `get_reg` I think it'd probably be better to use a single function definition but cfg the unsafe block since that's the only difference between the functions

Reply:
Agreed! Originally split out due to `error[E0658]: attributes on expressions are experimental` but `cfg`-ing the `unsafe` blocks is a clean workaround, thanks!